### PR TITLE
[serve] Add delete_endpoint call

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

If I can create an endpoint, I should be able to delete it!

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/serve/master.py [hunk 1]
```diff
@@ -479,6 +493,47 @@ async def create_endpoint(self, route, endpoint, methods):
...
+                    route_to_delete = route
+                    break
+            else:
+                # This should never happen, we either add to or delete from
+                # both self.traffic_policies and self.routes.
+                assert False, "No route found for endpoint '{}'".format(
+                    endpoint)
+
+            # Remove the routing entry.
+            del self.routes[route_to_delete]
```
Comment (simon-mo) [inline python/ray/serve/master.py:523]:
```suggestion
            for route, (route_endpoint, _) in self.routes.items():
                if route_endpoint == endpoint:
                    route_to_delete = route
                    del self.routes[route_to_delete]
```
Just making this no op if endpoint doesn't exist in self.routes?

File: python/ray/serve/tests/test_api.py [hunk 2]
```diff
@@ -306,3 +306,28 @@ def function2():
...
+    serve.set_traffic("delete_endpoint", {"delete-endpoint:v1": 1.0})
+
+    assert requests.get(
+        "http://127.0.0.1:8000/delete-endpoint").text == "hello"
+
+    # Check that deleting the endpoint doesn't delete the backend.
+    serve.delete_endpoint("delete_endpoint")
+    serve.create_endpoint("delete_endpoint", "/delete-endpoint")
+    serve.set_traffic("delete_endpoint", {"delete-endpoint:v1": 1.0})
+
```
Comment (simon-mo) [inline python/ray/serve/tests/test_api.py:331]: test a no route endpoint as well?
```
serve.create_endpoint("no-route") # no http routes
...
```

File: python/ray/serve/master.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/master.py:523]: This is an internal invariant check, see the comment

Comment (simon-mo) [reply python/ray/serve/master.py:523]: The invariant might not be true for no route endpoints?

**Review by simon-mo — APPROVED** [review]
LGTM as long as this works for no route endpoint.

**Pull request merged**